File: drivermonitoring/enviroment.py
import cv2
import time
import numpy as np

# main class
# we need to use numpy to optimize the code, // learn opencv optimization
class enviroment:

    def __init__(self):
        
        # exposure values
        self._no_exposure = 0
        self._little_exposure = 65
        self._slight_exposure = 130
        self._normal_exposure = 194
        self._over_exposure = 255

        # saturation values
        self._no_saturation = 0
        self._little_saturation = 65
        self._slight_saturation = 130
        self._normal_saturation = 194
        self._over_saturation = 255

        #frames per second values
        self.prev_frame_time = 0.0
        self.new_frame_time = 0.0


    def brightness_level(self, frame: np.ndarray) -> str:

        """ 
        parameters
        _________

        Frame : we take the frames as the input

        function
        ________

        we take frame and convert to HSV color space. We get the last V value and calculate the mean of the value.

        using the mean we determine the level of brightness : 1. no brightness, 2. little brightness 3. Normal brightness 4. Over brightness
        
        return
        _______

        str : the output string of the levels of brightness

        """

        hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
        value = hsv[...,2].mean()
        response = []

        if value == self._no_exposure:
            response.append('No exposure')
    
        if value > self._no_exposure and value <self._little_exposure:
            response.append('Minimal exposure')

        if value > self._little_exposure  and value < self._slight_exposure:
            response.append('slight exposure')

        if value > self._slight_exposure  and value < self._normal_exposure:
            response.append('Normal exposure')

        if value > self._normal_exposure  and value < self._over_exposure:
            response.append('over exposure')

        return response[0], value

    def saturation_level(self, frame: np.ndarray) ->str:

        """ 
        parameters
        _________

        Frame : we take the frames as the input

        function
        ________

        we take frame and convert to HSV color space. We get the last S value and calculate the mean of the value.

        using the mean we determine the level of brightness : 1. no staturation, 2. little staturation 3. Normal staturation 4. Over staturation
        
        return
        _______

        str : the output string of the levels of saturation

        """

        hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
        value = hsv[...,1].mean()
        response = []

        if value == self._no_saturation:
            response.append('No saturation')
    
        if value > self._no_saturation and value <self._little_saturation:
            response.append('Minimal saturation')

        if value > self._little_saturation  and value < self._slight_saturation:
            response.append('slight saturation')

        if value > self._slight_saturation  and value < self._normal_saturation:
            response.append('Normal saturation')

        if value > self._normal_saturation  and value < self._over_saturation:
            response.append('over saturation')

        return response[0], value

    def frames_per_second(self) ->int:

        """ 
        parameters
        _________

        None

        function
        ________

        we determine the number of frames per second of the frames

        
        return
        _______

        int : the number of frames per second

        """

        self.new_frame_time = time.time()

        fps = 1/(self.new_frame_time-self.prev_frame_time)

        self.prev_frame_time = self.new_frame_time
        fps = int(fps)

        return fps

File: drivermonitoring/cpp/src/src/enviroment.cpp
// enviroment file
// main class
    // check brightness level (more bugs)
    // check saturation level (some bugs)
    // get frames per second (done)

#include "enviroment.h";

saturation_R enviroment::saturation_level(Mat frame){
    // we calculate the mean of hsv[...,2]
    // returns two variables(string , float value)
    cvtColor(frame, hsvImage, COLOR_BGR2HSV);
    float value_S = (float)mean(hsvImage)[1];
    if(value_S == enviroment::_no_saturation){
        return { "no saturation", value_S };
    }
    else if(value_S > enviroment::_no_saturation && value_S < enviroment::_little_saturation){
        return { "Minimal saturation", value_S };
    }
    else if(value_S > enviroment::_little_saturation && value_S < enviroment::_slight_saturation){
        return { "slight saturation", value_S };
    }
    else if(value_S == enviroment::_slight_saturation && value_S < enviroment::_normal_saturation){
        return { "normal saturation", value_S };
    }
    else (value_S == enviroment::_normal_saturation && value_S < enviroment::_over_saturation);{
        return { "over saturation", value_S };
    }
};

brightness_R enviroment::brightness_level(Mat frame) {
    // we calculate the mean of hsv[...,1]
    // returns two variables(string , float value)
    cvtColor(frame, hsvImage, COLOR_BGR2HSV);
    float value = (float)mean(hsvImage)[2];
    if (value == enviroment::_no_exposure) {
        return { "no saturation", value };
    }
    else if (value > enviroment::_no_exposure && value < enviroment::_little_exposure) {
        return {"Minimal exposure", value};
    }
    else if (value > enviroment::_little_exposure && value < enviroment::_slight_exposure) {
        return {"slight exposure", value};
    }
    else if (value == enviroment::_slight_exposure && value < enviroment::_normal_exposure) {
        return {"normal exposure", value};
    }
    else (value == enviroment::_normal_exposure && value < enviroment::_over_exposure); {
        return {"over exposure", value};
    }
};


// we debug the enviroment class
void main() {
   enviroment env; // class declaration
   const string path = "E:\\surge\\resources\\video1.mp4";
   VideoCapture cap(path);
   Mat frame;
   while (true) {
       cap.read(frame);
       resize(frame, frame, Size(), 0.5, 0.5);
       double fps = cap.get(CAP_PROP_FPS);
       //cout << "Frames per second: " << fps << endl;
       cout << env.brightness_level(frame).value << endl;
       imshow("video", frame);
       waitKey(1);
   };
}